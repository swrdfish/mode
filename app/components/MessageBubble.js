import React from 'react'
import AutoLinker from 'autolinker'
import escapeHTML from 'escape-html'


class MessageBubble extends React.Component {
    constructor(props) {
        super(props)
    }

    render() {
        let item = this.props.item
        let className = item.isMine?'message-bubble mine':'message-bubble'
        let downloadButtonClass = item.isMine?"":"fa fa-download"
        let pretext = item.isMine?"Sent ":"Download "
        let text

        let itemRendered
        if(item.type == "file") {
            if(item.body.count > 1) {
                text = (
                    <a href="#" onClick={this.props.onDownload} data-id={item.body.id}>
                        {item.body.count} files&nbsp;&nbsp;&nbsp;
                        <span className={downloadButtonClass + " download-button"}/>
                    </a>
                )
            } else {
                text = (
                    <a href="#" onClick={this.props.onDownload} data-id={item.body.id}>
                        {item.body.filenames[0]}&nbsp;&nbsp;&nbsp;
                        <span className={downloadButtonClass + " download-button"}/>
                    </a>
                )
            }

            if(item.isMine) {
                text = "" + item.body.count
                text += item.body.count > 1?" files":" file"
            }
            className += " file"
            itemRendered = (
                <div className='message-bubble-wrapper'>
                    <div className={className}>
                        <span>
                            {pretext}&nbsp;&nbsp;{text}               
                        </span>
                        <span className='time-stamp'>{item.timeStamp.toLocaleString('en-US', { hour: 'numeric', minute: 'numeric', hour12: true })}</span>
                    </div>
                </div>
            )
        } else {
            let messageBody = AutoLinker.link(escapeHTML(item.body), {className: "message-bubble-link"})
            itemRendered = (
                <div className='message-bubble-wrapper'>
                    <div className={className}>
                        <span className='message-bubble-body' dangerouslySetInnerHTML={{ __html: messageBody }}></span>
                        <span className='time-stamp'>{item.timeStamp.toLocaleString('en-US', { hour: 'numeric', minute: 'numeric', hour12: true })}</span>
                    </div>
                </div>
            )
        }
        return itemRendered
    }
}

export default MessageBubble