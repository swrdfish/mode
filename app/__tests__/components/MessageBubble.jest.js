import React from 'react'
import { shallow, mount } from 'enzyme'
import MessageBubble from '../../components/MessageBubble'
import renderer from 'react-test-renderer';


Date.now = jest.fn(() => new Date(Date.UTC(2017, 7, 9, 8)).valueOf())

describe('Component: MessageBubble', () => {
    it('should render the message bubble', () => {
        let onDownload = jest.fn()
        let item = {
            type: "text",
            body: "hello world",
            id: 123,
            isMine: false,
            timeStamp: Date.now()
        }
        const wrapper = shallow(<MessageBubble item={item} onDownload={onDownload}/>)
        expect(wrapper.find(".message-bubble-wrapper")).toHaveLength(1)
    })

    it('should match previous snapshot', () => {
        let onDownload = jest.fn()
        let item = {
            type: "text",
            body: "hello world",
            id: 123,
            isMine: false,
            timeStamp: Date.now()
        }
        const tree = renderer
            .create(<MessageBubble item={item} onDownload={onDownload} />).toJSON();
        expect(tree).toMatchSnapshot();
    })

    it('should show links properly', () => {
        let onDownload = jest.fn()
        let item = {
            type: "text",
            body: "hello world <div class=\"xss-vulnerability\" style=\"color: red\">problem!</div> http://www.google.com",
            id: 123,
            isMine: false,
            timeStamp: Date.now()
        }
        const wrapper = mount(<MessageBubble item={item} onDownload={onDownload}/>)

        expect(wrapper.find(".message-bubble-wrapper")).toHaveLength(1)
        expect(wrapper.find(".message-bubble")).toHaveLength(1)
        expect(wrapper.find(".message-bubble-body")).toHaveLength(1)
        let messageBody = wrapper.find(".message-bubble-body").render()
        expect(messageBody.find(".xss-vulnerability")).toHaveLength(0)
        expect(messageBody.find(".message-bubble-link")).toHaveLength(1)
    })

    it('should render download links properly', () => {
        let onDownload = jest.fn()
        let item = {
            type: "file",
            body: {
                count: 1,
                filenames: ["somefile.ext"]
            },
            id: 123,
            isMine: false,
            timeStamp: Date.now()
        }

        const wrapper = mount(<MessageBubble item={item} onDownload={onDownload}/>)
        expect(wrapper.find(".file")).toHaveLength(1)
        expect(wrapper.find(".download-button")).toHaveLength(1)
        wrapper.find(".download-button").simulate('click')
        expect(onDownload).toHaveBeenCalled()
        expect(wrapper.find(".file").html().includes("Download")).toBeTruthy()
    })

    it('should render download link for multiple files', () => {
        let onDownload = jest.fn()
        let item = {
            type: "file",
            body: {
                count: 5,
                filenames: ["somefile1.ext", "somefile2.ext", "somefile3.ext", "somefile4.ext", "somefile5.ext"]
            },
            id: 123,
            isMine: false,
            timeStamp: Date.now()
        }

        const wrapper = mount(<MessageBubble item={item} onDownload={onDownload}/>)
        expect(wrapper.find(".file")).toHaveLength(1)
        expect(wrapper.find(".download-button")).toHaveLength(1)
        wrapper.find(".download-button").simulate('click')
        expect(onDownload).toHaveBeenCalled()
        expect(wrapper.find(".file").html().includes("files")).toBeTruthy()  
    })

    it('should display proper sent status', () => {
        let onDownload = jest.fn()
        let item = {
            type: "file",
            body: {
                count: 5,
                filenames: ["somefile1.ext", "somefile2.ext", "somefile3.ext", "somefile4.ext", "somefile5.ext"]
            },
            id: 123,
            isMine: true,
            timeStamp: Date.now()
        }

        const wrapper = mount(<MessageBubble item={item} onDownload={onDownload}/>)
        expect(wrapper.find(".file")).toHaveLength(1)
        expect(wrapper.find(".download-button")).toHaveLength(0)
        expect(wrapper.find(".file").html().includes("Sent")).toBeTruthy()        
    })
})