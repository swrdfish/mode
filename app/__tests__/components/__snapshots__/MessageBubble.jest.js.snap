// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: MessageBubble should match previous snapshot 1`] = `
<div
  className="message-bubble-wrapper"
>
  <div
    className="message-bubble"
  >
    <span
      className="message-bubble-body"
      dangerouslySetInnerHTML={
        Object {
          "__html": "hello world",
        }
      }
    />
    <span
      className="time-stamp"
    >
      1,502,265,600,000
    </span>
  </div>
</div>
`;
